### PR TITLE
[homekit] Added support for contact sensors

### PR DESCRIPTION
**Pull request by philipparndt**
Signed-off-by: Philipp Arndt <2f.mail@gmx.de> (github: philipparndt)

With this change it is possible to expose contact sensors to HomeKit.
Within the Home application it is possible to change the type of the sensor to
- garage door opener
- contact sensor
- window
- door
- jalousie

The reason to introduce this and not to use the default switch is, that the status of the contact sensor must be manually changed within the home application.

The configuration is simple and follows the rules of the current homekit add-on. A item can be marked as a contact sensor by adding the "ContactSensor" tag.

With this PR I've updated `HAP-Java` to the current release 1.1.4 (see https://github.com/beowulfe/HAP-Java/compare/hap-1.1.3...hap-1.1.4)

I took the build from https://search.maven.org/artifact/com.beowulfe.hap/hap/1.1.4/jar
this is necessary to add support for more accessories (see also https://github.com/openhab/openhab2-addons/pull/3487)

### CODE REVIEW

**Comment by zbot473** [comment]
Is this possible to test this on your own installation?

**Comment by frank-f** [comment]
I'd love to see this included in the official add-on, but I think this needs some little tweaks.
OpenHAB has an Item type for contact sensors (Contact) and I think for consistencies sake it would make a lot of sense to use it instead of the Switch type.
I have modified the code like that, tested it on my setup and it works nicely.
I don't know how to attach my changes to this PR. I can create a new one tho if that helps. Just drop me a line here in that case.

**Comment by zbot473** [comment]
@frank-f You can't change a pr that you don't own. Can you send me the jar that you used on your installation?

**Comment by philipparndt** [comment]
Thank you for the feedback. I've changed the SwitchItem to a ContactItem, this is much better. I didn't know that type as I'm new to openHUB.

I've tested a snapshot build on my installation.

**Comment by frank-f** [comment]
> Thank you for the feedback. I've changed the SwitchItem to a ContactItem, this is much better. I didn't know that type as I'm new to openHUB.

I noticed two more things. The CI build fails because openhab2-addons/addons/io/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitContactSensorImpl.java still says 2018. The most recent master branch requires all files to say 2019 and you should update openhab2-addons/addons/io/org.openhab.io.homekit/README.md to say Contact instead of Switch. It would confuse users otherwise.

**Comment by frank-f** [comment]
> @frank-f Can you send me the jar that you used on your installation?

Just clone the git repo, go to the binding directory and type "mvn install". It will build a SNAPSHOT jar from it and tell you the location towards the end.

**Comment by zbot473** [comment]
@frank-f where do i clone it? in the addons dir or anywhere then move the jar

**Comment by philipparndt** [comment]
thank you for the input, I made the changes on copyright date and documentation.

**Comment by zbot473** [comment]
@philipparndt How do i build it locally?

**Comment by zbot473** [comment]
I built it and put the jar in the addons directory in openhab2 but it doesn't show up

**Comment by frank-f** [comment]
> I built it and put the jar in the addons directory in openhab2 but it doesn't show up
> When i try to run the jar, it can't find the manifest. I don't know if this is related

You will not see the binding in paper ui because it was not installed that way.
If you're sure it's missing, open the openHAB console and type: bundle:start org.openhab.io.homekit

**Comment by zbot473** [comment]
@philipparndt It works on my Mac, but i'll have to test it out on a lower power device such as a RasPi, (it will probably run just fine). One thing is that if the item state is "NULL", it shows open. I think you can send a Status fault to 1 when the item state is "NULL" as described in the HAP-Specification. Great work though!

**Comment by zbot473** [comment]
Im going to make a topic on the openHAB community to hopefully get this integrated

**Comment by martinvw** [comment]
@beowulfe looks fine to me, could you do a quick check? Thanks!

File: addons/io/org.openhab.io.homekit/build.properties [hunk 1]
```diff
@@ -13,3 +13,4 @@ bin.includes = META-INF/,\
                lib/org.zeromq.curve25519-java-0.1.0.jar,\
                about.html,\
                ESH-INF/
+
```
Comment (Hilbrand) [inline addons/io/org.openhab.io.homekit/build.properties]: trailing spaces/tabs were added here.

File: addons/io/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitContactSensorImpl.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import org.openhab.io.homekit.internal.HomekitTaggedItem;
+
+import com.beowulfe.hap.HomekitCharacteristicChangeCallback;
+import com.beowulfe.hap.accessories.ContactSensor;
+import com.beowulfe.hap.accessories.properties.ContactState;
+
+/**
+ * Implements Contact sensor using an Item that provides an On/Off state.
+ *
+ * @author Philipp Arndt
```
Comment (Hilbrand) [inline addons/io/org.openhab.io.homekit/src/main/java/org/openhab/io/homekit/internal/accessories/HomekitContactSensorImpl.java]:
```suggestion
 * @author Philipp Arndt - Initial contribution
```

**Comment by wborn** [comment]
Can you also resolve the merge conflicts? 

> With this PR I've updated HAP-Java to the current release 1.1.4 (see beowulfe/HAP-Java@hap-1.1.3...hap-1.1.4)

It was already updated with https://github.com/openhab/openhab2-addons/pull/4535. So if you rebase on the latest master these changes no longer need to be part of this PR.

**Comment by timcharper** [comment]
@wborn  @philipparndt FYI I've been doing a lot of work on the plugin that I intend to work to integrate soon.

Me and @ccutrer have been working on getting some changes merged upstream in to HAP-Java; I've been waiting for that to stabilize before getting some patches merged in

https://github.com/timcharper/openhab2-addons/tree/2.4.0/homekit

I had already added support for contact sensors but I can easily just rebase on top of this and discard what I've done, (except for perhaps the part about also supporting switch items as a contact sensor accessory since some z-wave devices report contact as switches apparently?)

**Comment by ccutrer** [comment]
FYI I've already "rebased" (I had to cherry-pick one by one cause git was really mad about rebasing between 2.4.0 and master) mine and @timcharper's combined HomeKit branch on top of current openhab2 master: https://github.com/ccutrer/openhab2-addons/tree/homekit. For the resolution of ContactItemImpl, I basically took all of @timcharper , since it supports both SwitchItem and ContactItem, and also battery support (the latter will likely be refactored again, pending resolution of https://github.com/beowulfe/HAP-Java/issues/63 ).

**Comment by timcharper** [comment]
@ccutrer I think I'll go ahead and take your rebase; I'm thinking that with the preparation to integrate changes in to OpenHab2 it's probably time to base all work off of master

**Pull request merged**